File: 02 백준/02 구현/Python/Prob20546.py

"""
title  : 20546. 기적의 매매법 (Silver 5)
time   : 40ms      120ms
memory : 31256KB   113112KB
"""

Money = int(input())                                # 초기 현금
MachineDuck = list(map(int, input().split()))       # 1일부터 14일까지의 주가

# 준현이
money_bnp = Money   
count_bnp = 0

# 성민이
money_timing = Money
count_timing = 0

for num, price in enumerate(MachineDuck):

    # 준현
    count_bnp += (money_bnp // price)
    money_bnp %= price

    # 성민
    temp = MachineDuck[num:num+4]

    if len(temp) < 4:
        continue

    if temp[0] < temp[1] < temp[2] < temp[3] and count_timing > 0:         # 매도
        money_timing += (count_timing * temp[3])
        count_timing = 0
    elif temp[0] > temp[1] > temp[2] > temp[3]:                             # 매수
        count_timing += (money_timing // temp[3])
        money_timing %= temp[3]

answer_bnp = money_bnp + count_bnp * MachineDuck[-1]
answer_timing = money_timing + count_timing * MachineDuck[-1] 

if answer_bnp < answer_timing:
    print("TIMING")
elif answer_bnp > answer_timing:
    print("BNP")
else:
    print("SAMESAME")
    

File: 02 백준/02 구현/Java/src/Prob20546.java

/**
 * title  : 20546. 기적의 매매법 (Silver 5)
 * time   : 100ms
 * memory : 14240KB
 */

import java.io.*;
import java.util.*;

public class Prob20546 {

    public static void main(String[] args) throws IOException {

        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int Length = 14;

        int money =  Integer.parseInt(br.readLine());
        int[] price = new int[Length];
        
        StringTokenizer st = new StringTokenizer(br.readLine());
        for (int i = 0; i < 14; i++) {
            price[i] = Integer.parseInt(st.nextToken());
        }

        int amoney = money;
        int acount = 0;

        int bmoney = money;
        int bcount = 0;

        for (int p : price) {
            acount += (amoney / p);
            amoney %= p;
        }

        for (int i = 3; i < Length; i++) {

            if (price[i-3] < price[i-2] && price[i-2] < price[i-1] && price[i-1] < price[i]) {
                bmoney += (price[i] * bcount);
                bcount = 0;
            } else if (price[i-3] > price[i-2] && price[i-2] > price[i-1] && price[i-1] > price[i]) {
                bcount += (bmoney / price[i]);
                bmoney %= price[i];
            }
        }

        int aanswer = amoney + acount * price[Length-1];
        int banswer = bmoney + bcount * price[Length-1];

        if (aanswer == banswer) System.out.println("SAMESAME");
        else if (aanswer < banswer) System.out.println("TIMING");
        else System.out.println("BNP");
    }
}